### PR TITLE
Change categorical inference to look at sample of 10K rows and 80% non-unique

### LINKED ISSUE #1051: Change categorical inference to look at sample of 10K rows and 80% non-unique
- As a user, I wish I could use Woodwork to infer Categorical Logical Type. 
- This inference function would be based on a 10K sample, and 80% of the rows being non-unique. 
  - (Rather than the current implementation which looks at string length).

### PR DESCRIPTION
**Pull request by davesque**
- Attempts to make inference of categorical types more consistent.
- Closes #1051 

This is really just a first stab at this.  I ended up keeping the `numerical_categorical_threshold` setting since it appears that it was being used in a couple places (and to make inference of categoricals from numericals optional).  I ended up having to update a lot of tests that were broken by this update.

### CODE REVIEW

**Comment by davesque** [comment]
Still need to update release notes.

File: woodwork/tests/accessor/test_statistics.py [hunk 1]
```diff
@@ -80,7 +80,8 @@ def test_accessor_replace_nans_for_mutual_info():
         'str_no_nan': pd.Series(['test', 'test2', 'test2', 'test']),
         'dates': pd.Series(['2020-01-01', None, '2020-01-02', '2020-01-03'])
     })
-    df_nans.ww.init()
+    df_nans.ww.init(logical_types={"bools": "Categorical", "str":
+                                   "Categorical", "str_no_nan": "Categorical"})
```
Comment (thehomebrewnerd) [inline woodwork/tests/accessor/test_statistics.py:84]: Is this because we would now infer these column types differently than we did before?

File: docs/source/release_notes.rst [hunk 2]
```diff
@@ -17,6 +17,7 @@ Future Release
         * Entirely null columns are now inferred as the Unknown logical type (:pr:`1043`)
         * Add helper functions that check for whether an object is a koalas/dask series or dataframe (:pr:`1055`)
         * ``TableAccessor.select`` method will now maintain dataframe column ordering in TableSchema columns (:pr:`1052`)
+        * The criteria for categorical type inference have changed (:pr:`1065`)
```
[the remark this answers is not in the document]
Comment (davesque) [reply docs/source/release_notes.rst:20]: Ah, yeah you're right.  Didn't think of that.

File: woodwork/tests/accessor/test_statistics.py (hunk 1 above)
Comment (davesque) [reply woodwork/tests/accessor/test_statistics.py:84]: Yeah, it's because `_replace_nans_for_mutual_info` has some if conditionals that include things like `column.is_numeric` and `column.is_categorical`.  Those end up taking different branches after this change because the columns mentioned in the line you highlight are no longer inferred as categorical.  Updating the sample data to be inferred differently felt like a pain in the butt because I'd also have to update all the lines further down in the test case.  On the other hand, all we really care about testing is that the internal logic of the `_replace_nans_for_mutual_info` works correctly.  It seemed just as effective to do that by forcing certain logical types with `ww.init(logical_types=...)` than by having those types inferred.

File: woodwork/type_sys/inference_functions.py [hunk 3]
```diff
@@ -17,26 +17,22 @@ def get_inference_sample(series: pd.Series) -> pd.Series:
...
+        categorical_threshold = ww.config.get_option('categorical_threshold')
+
+        return _is_categorical(sample, categorical_threshold)
+
+    if pdtypes.is_float_dtype(series.dtype) or pdtypes.is_integer_dtype(series.dtype):
+        numeric_categorical_threshold = ww.config.get_option('numeric_categorical_threshold')
+        if numeric_categorical_threshold is not None:
+            return _is_categorical(series, numeric_categorical_threshold)
+        else:
+            return False
```
[the remark this answers is not in the document]
Comment (davesque) [reply woodwork/type_sys/inference_functions.py:34]: @tamargrey I think the overall way I was trying to structure things was so that the top-level if statements are trying to identify the overall category of Pandas dtype that a series might fall into (explicitly categorical?  some kind of object?  numerical?).  Then, within each of those if blocks, the function should figure out how to explicitly say whether or not it cares about the series and give a concrete answer.  From that, I think it's a *little* easier to follow the possible execution of the function in your head.  The reader of the function sees an if block that has a return statement in the positive case and also in the negative case and knows without any further examination that the function will return at that point.  Otherwise, they might have to read further down to figure out exactly what will happen.  We could save a line of code by not including the else block but I think we lose a little in readability, especially if the function gets updated later to include further logic before the final `return False` at the bottom of the function.

File: woodwork/type_sys/utils.py [hunk 4]
```diff
@@ -141,3 +141,32 @@ def _get_specified_ltype_params(ltype):
         # Do not reveal parameters for an uninstantiated LogicalType
         return {}
     return ltype.__dict__
+
+
+def _is_categorical(series: pd.Series, threshold: float) -> bool:
```
Comment (rwedge) [inline woodwork/type_sys/utils.py:146]: should we call this `_is_categorical_series` similar to `_is_numeric_series` in the same file

might help differentiate from `ColumnSchema.is_categorical`

File: woodwork/type_sys/inference_functions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tamargrey) [reply woodwork/type_sys/inference_functions.py:34]: Thanks for the explanation! That makes sense to me, happy to keep as is

File: woodwork/type_sys/utils.py [hunk 5]
```diff
@@ -141,3 +141,32 @@ def _get_specified_ltype_params(ltype):
...
+        # doesn't provide an API that would give you any indication ahead of
+        # time if a series with object dtype has any unhashable elements.
+        if "unhashable type" in e.args[0]:
+            return False
+        else:
+            raise  # pragma: no cover
+    count = series.count()
+
+    if nunique == 0 or count == 0:
+        return False
```
Comment (rwedge) [inline woodwork/type_sys/utils.py:168]: if nunique is 0 we could exit before doing a count

File: woodwork/type_sys/inference_functions.py [hunk 6]
```diff
@@ -17,26 +17,22 @@ def get_inference_sample(series: pd.Series) -> pd.Series:
 
 
 def categorical_func(series):
-    categorical_threshold = ww.config.get_option('categorical_threshold')
-    numeric_categorical_threshold = ww.config.get_option('numeric_categorical_threshold')
+    if pdtypes.is_categorical_dtype(series.dtype):
+        return True
```
[the remark this answers is not in the document]
Comment (davesque) [reply woodwork/type_sys/inference_functions.py:21]: > Do we have specific tests that confirm that he category dtype will end up with Categorical logical type in situations where Categorical is not inferred otherwise?

Not sure.  I suppose that whole edge case depends on behavior defined in the calling context of `categorical_func` which is `TypeSystem.infer_logical_type`.  Are you saying that you think we need a test to touch the edge case of when a series has the categorical dtype but might possibly still be inferred as some other type by the logic in `TypeSystem.infer_logical_type`?

File: woodwork/type_sys/utils.py (hunk 4 above)
Comment (davesque) [reply woodwork/type_sys/utils.py:146]: Yeah, seems like a good idea.

File: woodwork/type_sys/utils.py (hunk 5 above)
Comment (davesque) [reply woodwork/type_sys/utils.py:168]: Yeah, and now that you mention it I think nunique == 0 implies count == 0.

File: woodwork/type_sys/inference_functions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tamargrey) [reply woodwork/type_sys/inference_functions.py:21]: I'm more thinking about the opposite situation, I think, in which a column is below the categorical threshold but has the category dtype so it's still `Categorical` logical type. Just would want to confirm that behavior. 

But I think the edge case you brought up is also relevant, though it might be maybe something like a column of emails with dtype category. If the email inference recognizes the email values, then the logical type should be email since it's a more specific type. But that's more of a type system test than a ltype inference test.

[the remark this answers is not in the document]
Comment (davesque) [reply woodwork/type_sys/inference_functions.py:21]: > But that's more of a type system test than a ltype inference test.

Yup, agreed.

> a column is below the categorical threshold but has the category dtype so it's still Categorical logical type

Sure, I guess we can test for that if it seems like an important property we want our system to have.  And, just to confirm, this would be the correct behavior, right?  If it's explicitly marked as being categorical by Pandas (by possessing the categorical dtype), then it should also have the Categorical logical type?

[the remark this answers is not in the document]
Comment (tamargrey) [reply woodwork/type_sys/inference_functions.py:21]: (realize I meant _above_ the categorical threshold in the comment above, which reminds me that we should update the `setting_config_options` guide as well here).

I think this behavior makes sense, though probably worth getting input from @gsheni  and @thehomebrewnerd to see if they agree. Though, it's worth noting that there will not be an equivalent in Koalas, which has no `category` dtype, so the `string` dtype gets used.

[the remark this answers is not in the document]
Comment (davesque) [reply woodwork/type_sys/inference_functions.py:21]: > realize I meant *above* the categorical threshold in the comment above

Yep, assumed that's what you meant.

File: docs/source/release_notes.rst [hunk 7]
```diff
@@ -32,6 +32,10 @@ v0.5.1 Jul 22, 2021
     Thanks to the following people for contributing to this release:
     :user:`davesque`, :user:`frances-h`, :user:`jeff-hernandez`, :user:`simha104`, :user:`tamargrey`, :user:`thehomebrewnerd`
 
+Breaking Changes
+++++++++++++++++
+    * The criteria for categorical type inference have changed (:pr:`1065`)
```
Comment (rwedge) [inline docs/source/release_notes.rst:37]: Usually for breaking changes we put the one-liner with the PR number in the Changes section above and in Breaking Changes provide more details about how the change works / will impact users

File: woodwork/type_sys/inference_functions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (thehomebrewnerd) [reply woodwork/type_sys/inference_functions.py:21]: I think it makes sense that a column with a `category` dtype would get inferred as a `Categorical` column in Woodwork, even if its values would cause it to be inferred as something else based on the threshold approach.

File: docs/source/release_notes.rst (hunk 7 above)
Comment (davesque) [reply docs/source/release_notes.rst]: https://github.com/alteryx/woodwork/pull/1065/commits/0b346dad2e24e6ea39f27aa8db187a20b5f63574

File: woodwork/type_sys/inference_functions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (davesque) [reply woodwork/type_sys/inference_functions.py:21]: Recommended change made in this commit: https://github.com/alteryx/woodwork/pull/1065/commits/cd3456a8e589637f4fede17fd72c9934679d1a78

**Review by rwedge — APPROVED** [review]
Looks good - assuming no more requested changes from @thehomebrewnerd and @tamargrey

File: woodwork/tests/type_system/test_ltype_inference.py [hunk 8]
```diff
@@ -114,8 +114,21 @@ def test_categorical_inference(categories):
             assert isinstance(inferred_type, Categorical)
 
 
+def test_categorical_inference_based_on_dtype(categories_dtype):
+    """
+    This test specifically targets the case in which a series can be inferred
+    to be categorical strictly from its pandas dtype, but would otherwise be
+    inferred as some other type.
+    """
```
Comment (thehomebrewnerd) [inline woodwork/tests/type_system/test_ltype_inference.py:122]: I like this explanation - sometimes think it would be nice for all of our tests to have this so we are clear what they are testing months from now...

**Review by thehomebrewnerd — APPROVED** [review]
LGTM!

**Review by tamargrey — APPROVED** [review]
lgtm! Noticed a tiny typo that's probably been in the docs for a while, but other than that, looks great

File: docs/source/guides/setting_config_options.ipynb [hunk 9]
```diff
@@ -29,13 +29,13 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "The output of `ww.config` lists each of the available config variables followed by it's current setting. In the output above, the `categorical_threshold` config variable has been set to `10` and the `numeric_categorical_threshold` has been set to `-1`. \n",
+    "The output of `ww.config` lists each of the available config variables followed by it's current setting. In the output above, the settings for the `categorical_threshold` and `numeric_categorical_threshold` config variables are visible.\n",
```
[the remark this answers is not in the document]
Comment (davesque) [reply docs/source/guides/setting_config_options.ipynb:32]: Yeah, one of my least favorite grammatical errors.  Thanks for catching it.
